src/app/CommandSender.cpp

```diff
@@ -61,10 +62,22 @@ CHIP_ERROR CommandSender::SendCommandRequest(NodeId aNodeId, Transport::AdminId
     VerifyOrExit(mpExchangeCtx != nullptr, err = CHIP_ERROR_NO_MEMORY);
     mpExchangeCtx->SetResponseTimeout(kImMessageTimeoutMsec);
 
-    err = mpExchangeCtx->SendMessage(
-        Protocols::InteractionModel::MsgType::InvokeCommandRequest, std::move(mCommandMessageBuf),
-        Messaging::SendFlags(Messaging::SendMessageFlags::kExpectResponse).Set(Messaging::SendMessageFlags::kNoAutoRequestAck));
-    SuccessOrExit(err);
+    event.Type                              = DeviceLayer::DeviceEventType::kInteractionModelCommandRequest;
+    event.InteractionModelEvent.ExchangeCtx = mpExchangeCtx;
+
+    // ChipDeviceEvent can only contains the structures with trivial copy assignment operator, so we can't use PacketBufferHandle
+    // and SendFlags directly within ChipDeviceEvent.
+    event.InteractionModelEvent.Payload   = std::move(mCommandMessageBuf).UnsafeRelease();
+    event.InteractionModelEvent.SendFlags = Messaging::SendFlags(Messaging::SendMessageFlags::kExpectResponse)
+                                                .Set(Messaging::SendMessageFlags::kNoAutoRequestAck)
+                                                .Raw();
+
+    // Currently, Interaction Model protocol is implemented in app thread instead of CHIP thread, we need to lock CHIP stack to
+    // prevent it from operating before we post an event to the event queue associated with the CHIP thread.
```

Comment:
Assumption of "app thread" versus "CHIP thread" may not hold over long term. This comment may get stale.

Reply:
Yes, this only valid with current IM implementation within client thread.